design/external-references.md

```diff
@@ -0,0 +1,394 @@
+- **REQ3**: Allow CDK apps to reference resources that were created elsewhere (by other CDK apps, manually or any other mean) by referencing the physical identity. For example, an S3 Bucket ARN should be sufficient in order to reference
+an existing S3 bucket.
+- **REQ4**: If possible, implement the mechanism such that it can be later used to publish and consume constructs through other key/value mechanism such as environment variables, SSM parameters, etc.
+- **REQ5**: If a resource is exported twice under the same export name, only a single set of outputs will be created ([#1496](https://github.com/awslabs/aws-cdk/issues/1496)).
+- **REQ6**: If the same resource is imported twice in the same stack, the "import" method will return the same object.
+- **REQ7**: For some resources (like VPC for example), it should be possible look up the resource in the current environment by querying through an environmental context provider.
+- **REQ8**: It should be possible to resolve import values either during synthesis time (i.e. via an environmental context provider) or during deploy time (i.e. via `Fn::ImportValue`). There are certain resources that will _require_ the use of synthesis-time resolution due to their complex representation.
+
+## Approach
+
+As usual, we'll take a layered approach.
+
+**Serializable constructs**: At the low-level, we'll define what it means for a construct to be "__serializable__" through string key/value context. The mechanism will be recursive such that it will be possible to write/read singular values and write/read serializable objects (which, themselves will write/read their own values and so forth).
+
+**Serialization context**: the notion of a "string key/value context" represents the lowest common denominator for serialization, and specifically is the only type supported by CloudFormation's cross-stack export/import mechanism (CloudFormation outputs can only be of "string" type). If a construct can serialize itself through a set of key/values, we can pass it via CloudFormation's import/export, environment variables, SSM parameters, etc.
+
+**CloudFormation imports/exports as a serialization context**: on top of that, we will implement a serialization/deserialization context based on AWS CloudFormation's import/exports mechanism. "writing" a value means defining an Output with an Export and "reading" a value means `Fn::ImportValue` with this export name. The "export name" will be used as a prefix that represent an object, and sub-objects will be serialized by adding another component to the export name.
+
+**Synthesis-time Imports**: we will also define an environmental context provider that will import values during synthesis and store them in `cdk.json`. This approach is more robust since it will allow CDK code to reason about concrete values instead of opaque tokens. It's especially needed for situations where a list of values needs to be passed and the arity is required.
```

Comment:
This has a caveat in case some of the properties that are exported are mutable - how do you make sure you're using the latest/current value appropriately? (I'm okay assuming that "replaces" cannot happen, because `Fn::Import` would protect against that)